src/azul/queues.py

```diff
@@ -273,34 +231,22 @@ def limit_timeout(timeout):
             logger.info('Message count history (most recent first) is %r.',
                         list(reversed(total_lengths)))
 
-            if len(total_lengths) == total_lengths.maxlen and all(n == 0 for n in total_lengths) == empty:
-                logger.info('The queues are at the desired level.')
-                break
+            if len(total_lengths) == total_lengths.maxlen:
+                cummdiff = sum(
+                    abs(first - second)
+                    for first, second in more_itertools.pairwise(total_lengths)
+                )
+                if cummdiff == 0:
+                    final_length = total_lengths[-1]
+                    logger.info('The queues have stabilized.')
+                    break
 
-            # When draining, determine timeout dynamically
-            if empty:
-                # Estimate number of bundles first, if necessary
-                if num_expected_bundles is None:
-                    num_notifications = queue_lengths[config.notifications_queue_name()]
-                    if num_bundles < num_notifications:
-                        num_bundles = num_notifications
-                        start_time = time.time()  # restart the timer
-                else:
-                    num_bundles = num_expected_bundles
-                # It takes approx. 6 seconds per worker to process a bundle
-                # FIXME: Temporarily doubling the time, but needs fine-tuning
-                #        https://github.com/DataBiosphere/azul/issues/2147
-                #        https://github.com/DataBiosphere/azul/issues/2189
-                timeout = limit_timeout(12 * num_bundles / config.indexer_concurrency)
-
-            # Do we have time left?
-            remaining_time = start_time + timeout - time.time()
-            if remaining_time <= 0 and len(total_lengths) == total_lengths.maxlen:
-                raise Exception('Timeout. The queues are NOT at the desired level.')
-            else:
-                logger.info('Waiting for up to %.2f seconds for %s queues to %s ...',
-                            remaining_time, len(queues), 'drain' if empty else 'fill')
-                time.sleep(sleep_time)
+            logger.info('Waiting for %s queues to stabilize...', len(queues))
```

Comment:
It looks like you can roll this to the top of the loop and eliminate the duplication. If not, see my comment on the log statement above.

Reply:
That wouldn't be equivalent since the loop would sleep _before_ subsequent log statements instead of after. Instead of `log, [check, log, sleep], [check, log, sleep] ...` it would be `[log, check, sleep], [log, check, sleep]...`